### ISSUE TITLE
unnecessary_fold: .any and .all may not always be safe replacements of .fold

### ISSUE THREAD

**Issue by ghost** [rust-lang/rust-clippy#3351]
Clippy version: 0.0.212
https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#unnecessary_fold

Both .any and .all are short-circuiting while .fold evaluates every element in the iterator, if the iterator incurs side effects these two methods will have different behavior than .fold.

For example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=893a485dc791c89d95d6ce6f744af9bf

**Comment by camsteffen** [comment]
I just ran into this.

I think the short-circuiting methods (`any` and `all`) should be removed from this lint. If they are still wanted, they should be moved to a pedantic lint with known false positives.

At a minimum, the warning message should have a note like "this may change semantics" or "this may change side-effects".

**Comment by camsteffen** [comment]
I suppose it may be possible to detect cases where there are side-effects?

**Comment by maartendeprez** [comment]
Hi, I also ran into this. Clippy suggests replacing `.fold(false, |a,b| a || b)` with `.any(...)`, but applying the suggested fix would lead to subtle bugs in my code because of side effects in the iterator. I think this lint (in its current version) should at least be `MaybeIncorrect` and explain the differences.

**Comment by falsetru** [comment]
I also ran into this.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c74eadf27c94c189d1649b7f7a4c187c

**Comment by dbeckwith** [comment]
Also just noticed this. I think it's good to mention `any` as a possible alternative to `fold` but it should be very clear about the difference in short-circuiting behavior.

**Comment by vilgotf** [comment]
I also ran into this within tokio: (https://github.com/tokio-rs/tokio/pull/7816)

```rust
self.storage
    .iter()
    // Any signal of this kind arrived since we checked last?
    .filter(|event_info| event_info.pending.swap(false, Ordering::SeqCst))
    // Ignore errors if there are no listeners
    .map(|event_info| event_info.tx.send(()).is_ok())
    .fold(false, |acc, success| acc || success)
```
